AGS02MA.cpp

```diff
@@ -205,17 +205,34 @@ uint32_t AGS02MA::readUGM3()
 }
 
 
-bool AGS02MA::zeroCalibration()
+
+bool AGS02MA::manualZeroCalibration(uint16_t value)
 {
   _buffer[0] = 0x00;
   _buffer[1] = 0x0C;
-  _buffer[2] = 0xFF;
-  _buffer[3] = 0xF3;
-  _buffer[4] = 0xFC;
+  _buffer[2] = (uint8_t) (value >> 8);
+  _buffer[3] = (uint8_t) value;
+  _buffer[4] = _CRC8(_buffer, 4);
   return _writeRegister(AGS02MA_CALIBRATION);
 }
 
 
+AGS02MA::ZeroCalibration AGS02MA::getZeroCalibration() {
+  AGS02MA::ZeroCalibration zc;
+  if (_readRegister(AGS02MA_CALIBRATION))
+  {
+    _error = AGS02MA_OK;
+    zc.status = _dataMSB(_buffer);
+    zc.value = _dataLSB(_buffer);
+    if (_CRC8(_buffer, 5) != 0)
+    {
+      _error = AGS02MA_ERROR_CRC;
+    }
+  }
+  return zc;
```

Comment:
if readRegister() fails ZC is not initialized and can have any value (as it is a local var from a function)
Furthermore the _error should be set - although yes it is set in _readregister()

So the user should always check _error to be sure that  ZC contains meaningful data. 
Should we extend the ZC with a "bool valid" to make it a bit easier?

If there is a CRC error, can we trust the ZC.status and ZC.value?

```cpp
AGS02MA::ZeroCalibration AGS02MA::getZeroCalibration() {
  AGS02MA::ZeroCalibration zc = {0, 0};    // or -1, -1 ?
  if (_readRegister(AGS02MA_CALIBRATION) == false)
  {
    _error = AGS02MA_ERROR_CALIBRATION ...
  }
  else if (_CRC8(_buffer, 5) != 0)
  {
    _error = AGS02MA_ERROR_CRC;
  }
  else
  {
    _error = AGS02MA_OK;
    zc.status = _dataMSB(_buffer);    // as _buffer is a local var I think it doesn't need to be a parameter ...
    zc.value = _dataLSB(_buffer);
  }
  return zc;
}
```

Reply:
All good points. Should we change the function signature to be similar to my register dumping example?

```cpp
int|bool getZeroCalibrationData(ZeroCalibrationData &data);
```

Such that return value makes errors explicit, having to initialize the struct yourself though.

Reply:
Think that is better
(I know a lot is a matter of taste,  I try to think how less experienced people can work with it, think this latter is the better one)

Reply:
Well even experienced people can easily make the mistake of not error handling if it's not very explicit.
Case in point, both updated examples don't handle read calibration data errors atm :joy:
I'll go with the bool return value to match other methods, and update accordingly.

Reply:
Now with 80% more error and 120% more handling!